File: Baekjoon/32177 - 에어드롭/32177.cpp
#include <iostream>
#include <vector>
using namespace std;

#define FASTIO cin.tie(NULL); cout.tie(NULL); ios::sync_with_stdio(false);

struct point {
    int x, y, version;
};

int n, k, t;
point puang;
vector<int> parent, picture;
vector<point> v;

int Find(int a) {
    if(a == parent[a]) return a;
    else return parent[a] = Find(parent[a]);
}

void Union(int a, int b) {
    a = Find(a);
    b = Find(b);

    if(a > b) parent[a] = b;
    else parent[b] = a;
}

int main() {
    FASTIO

    cin >> n >> k >> t;

    for(int i = 0; i <= n; i++) {
        parent.push_back(i);
    }

    cin >> puang.x >> puang.y >> puang.version;

    v.push_back({puang.x, puang.y, puang.version});

    for(int i = 1; i <= n; i++) {
        int temp_x, temp_y, temp_version, temp_p;
        cin >> temp_x >> temp_y >> temp_version >> temp_p;

        v.push_back({temp_x, temp_y, temp_version});

        if(temp_p == 1) {
            picture.push_back(i);
        }
    }

    for(int i = 0; i <= n; i++) {
        for(int j = i+1; j <= n; j++) {
            int distance = (v[i].x-v[j].x)*(v[i].x-v[j].x) + (v[i].y-v[j].y)*(v[i].y-v[j].y);

            if(distance <= k*k && abs(v[i].version-v[j].version) <= t) {
                Union(i, j);
            }
        }
    }

    int picture_size = picture.size();
    bool check = false;

    for(int i = 0; i < picture_size; i++) {
        if(Find(picture[i]) == Find(0)) {
            cout << picture[i] << " ";
            check = true;
        }
    }

    if(check == false) {
        cout << 0;
    }

    return 0;
}

File: solution/32xxx/32177 - 에어드롭/32177.py
import sys
input = sys.stdin.readline

def find(parent, a):
    if a != parent[a]:
        parent[a] = find(parent, parent[a])

    return parent[a]

def union(parent, a, b):
    a = find(parent, a)
    b = find(parent, b)

    if a > b:
        parent[a] = b
    else:
        parent[b] = a

n, k, t = map(int, input().split())

parent = []
for i in range(0, n+1):
    parent.append(i)

point = []
puang_x, puang_y, puang_version = map(int, input().split())
point.append([puang_x, puang_y, puang_version])

picture = []
for i in range(1, n+1):
    temp_x, temp_y, temp_version, temp_picture = map(int, input().split())
    point.append([temp_x, temp_y, temp_version])

    if temp_picture == 1:
        picture.append(i)

for i in range(0, n+1):
    for j in range(i+1, n+1):
        distance = (point[i][0]-point[j][0])*(point[i][0]-point[j][0]) + (point[i][1]-point[j][1])*(point[i][1]-point[j][1])

        if distance <= k*k and abs(point[i][2]-point[j][2]) <= t:
            union(parent, i, j)

check = False

for i in range(0, len(picture)):
    if(find(parent, parent[picture[i]]) == find(parent, 0)):
        print(picture[i], end=" ")
        check = True

if check == False:
    print(0)